### PR TITLE
N(ext) Runner + Job: prep work

### PR DESCRIPTION
**Pull request by clebergnu**
This contains a large number of fixes and prep work, before the final integration PR.  A few notes:

* The selftests for the TastStateMachine is *functional* and can work with other spawners.  I've used it successfully with the podman spawner, and maybe we could add a pre release test with it)
* The status repo is receiving new utility methods as the final monitoring/UI/ results aspects of the job integration are being developed.
* Because the BP003 describe the triage step, I've decided to include the core/avocado requirement checker here, as it gives a concrete example of triaging in the context of the spawners. 
* If reviewers are interested in seeing this prep work in use, this [commit](https://github.com/clebergnu/avocado/commit/8c7b763cb48f996e2e5fa8304fa7e10f27ea9434) should give an idea.  As the commit message itself explains, there's some work pending there, but it should be possible to see the TaskTateMachine, workers, spawners, status server and status repo in use.

### CODE REVIEW

File: avocado/core/plugin_interfaces.py [hunk 1]
```diff
@@ -266,9 +266,34 @@ class Spawner(Plugin):
...
+        """
+
+    @staticmethod
+    @abc.abstractmethod
+    async def check_task_requirements(task_info):
+        """Checks if the requirements described within a task are available.
+
+        :param task_info: wrapper for a Task with additional runtime information
+        :type task_info: :class:`avocado.core.task.info.TaskInfo`
+        """
```
Comment (pevogam) [inline avocado/core/plugin_interfaces.py:299]: Hi @clebergnu, how complete is the API at this stage? Do we have any other required methods for a spawner plugin specified in any of the latest blueprints?

File: avocado/core/requirements.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     'environment TEXT,'
+     'requirement_type TEXT,'
+     'requirement TEXT,'
+     'FOREIGN KEY(environment_type) REFERENCES environment(environment_type),'
+     'FOREIGN KEY(environment) REFERENCES environment(environment),'
+     'FOREIGN KEY(requirement_type) REFERENCES requirement_type(requirement_type)'
+     ')'),
+    ('CREATE UNIQUE INDEX IF NOT EXISTS requirement_idx ON requirement '
+     '(environment_type, environment, requirement_type, requirement)')
+]
```
Comment (beraldoleal) [inline avocado/core/requirements.py:50]: Based on this code and on this commit message I would choose something that uses more like an "MVC pattern" and also would try to avoid a DB scheme here.

We don't need to discuss the choice for SQLite here, but my suggestion would be to split this into a proper package (for instance `avocado.core.persistence.backends.sqlite`) with a very basic interface, so if we decide to change this later, we could do it easily. Also for better organization, avoiding specific SQLite code inside the `requirements` module.

File: avocado/plugins/spawners/podman.py [hunk 3]
```diff
@@ -129,3 +132,55 @@ async def wait_task(task_info):
...
+        result = stdout.rstrip().decode() == MAGIC
+        if result:
+            set_requirement_on_cache(env_type, env,
+                                     req_type, req)
+            return True
+        return False
+
+    @staticmethod
+    async def check_task_requirements(task_info):
+        runnable_requirements = task_info.task.runnable.requirements
```
Comment (beraldoleal) [inline avocado/plugins/spawners/podman.py:173]: Here is one example where task and task_info are combined. You are already passing the task obj inside the task_info. So this could be `task.runnable.requirements` ;)

File: avocado/core/task/statemachine.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            pending = any([self._requested, self._triaging,
+                           self._ready, self._started])
+        return not pending
+
+
+class Worker:
+
+    def __init__(self, task_state_machine, spawner,
+                 max_triaging=None, max_running=None, task_timeout=None):
+        self._tsm = task_state_machine
```
Comment (beraldoleal) [inline avocado/core/task/statemachine.py:52]: s/task_state_machine/state_machine/g ?

s/_tsm/_state_machine/g ?

File: avocado/core/magic.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat, Inc. 2020
+# Author: Cleber Rosa <crosa@redhat.com>
+
+
+#: Magic number used to check that an "avocado" command is indeed
+#: present and matches one's expectations
+MAGIC = "971dfda4234b6abbea36ed1e939664f45c0190b234918d627449c8b578f8072e"
```
[the remark this answers is not in the document]
Comment (pevogam) [reply avocado/core/magic.py:17]: Hmmm, I was wondering the same. And usage of MAGIC of any kind is generally better avoided in software patterns.

[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/magic.py:17]: The idea is we want to make sure that "our" Avocado is present.

For comparison purposes, our older `runner-{remote,vm,docker}` plugins, would run `avocado --version` and check if `Avocado \d\d\.\d` (or something like this) was present.  But there could be a different application with the same name installed (on PyPI alone we have quite a lot of similar named packages: https://pypi.org/project/avocado/, https://pypi.org/project/avocado-messenger-server/).

So, instead of doing more heuristics, such as trying to find the right packages or files, I thought a magic number would avoid any confusion.

[the remark this answers is not in the document]
Comment (pevogam) [reply avocado/core/magic.py:17]: I think looking for a binary or an importable module of a particular version the way it was done before is a much more standard approach to check if a given dependency is provided. In this case `avocado --version` should be good enough for most practical situations. Are there important use cases where we need to verify greater authenticity for the avocado code? I guess in such extreme cases the magic hex number will not work either.

[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/core/magic.py:17]: Yes, I understand but there is a lot of code here just to print a MAGIC number (plus a different subcommand).

So, my suggestion is to print a small hash along with `avocado --version` and keep things simples.

[the remark this answers is not in the document]
Comment (pevogam) [reply avocado/core/magic.py:17]: Btw, how is a small hash in addition to the version contributing more, i.e. how is the version by itself not unique enough to find the avocado dependency and a versioned one in particular?

File: avocado/core/plugin_interfaces.py (hunk 1 above)
Comment (clebergnu) [reply avocado/core/plugin_interfaces.py:299]: Hi @pevogam, I expect that this API (as in the end of this PR) will be the final shape for releases 81.0 and 82.0.  Unfortunately, none of the blueprints touched on the subject of spawners directly.  After 82.0, we're free to expand/improve this API, so I'd love to hear your ideas.

Comment (pevogam) [reply avocado/core/plugin_interfaces.py:299]: That's great news then! Yes, I think I have some ideas that I want to try out but was waiting for a green flag like this one. :green_circle:

File: selftests/unit/test_task_info.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+class Info(TestCase):
+
+    def setUp(self):
+        runnable = Runnable('noop', 'noop')
+        task = Task('1-noop', runnable)
+        self.task_info = TaskInfo(task)
+
+    def test_empty(self):
+        self.assertIsNone(self.task_info.status)
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply selftests/unit/test_task_info.py:15]: > IMO this usage is very strange. `task.task_info.status`, instead of a direct `task.status`. I understand that at some point tasks will not have all information, but as I said before in a previous review, from the OOP perspective there is no problem with this being part of `Task`. I can't see the reason for this separation.

Do you think a task should know that it's supposed to be started by a spawner? And that it should have a place for the spawner's own handling? I find this very wrong.  You mention the "OOP perspective", but composition is widely used in OOP. 

> 
> We are creating a new class because only 4 attributes that are related to a Task.

Right, and I think they're in the right place.  If you look at how a task *can be run*, say, `Task.run()` or `avocado-runner-* task-run ...` it needs none of the attributes.  So they're realted to a Task, but they're not part of a Task.

File: avocado/core/magic.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/core/magic.py:17]: @pevogam Yes, there are no advantages (unless I'm missing something) with the simple version. We could check if `version >= 80` for instance. So, I'm ok with a simple version too.

File: avocado/plugins/nrun.py [hunk 7]
```diff
@@ -151,11 +152,12 @@ def run(self, config):
                 LOG_UI.error("Spawner not implemented or invalid.")
                 sys.exit(exit_codes.AVOCADO_JOB_FAIL)
 
+            self.pending_tasks = [TaskInfo(task) for task in self.pending_tasks]
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/plugins/nrun.py]: I'm not following the question... the task is "saved" inside the `TaskInfo()` instance created.

File: avocado/core/requirements.py (hunk 2 above)
Comment (clebergnu) [reply avocado/core/requirements.py:50]: > Based on this code and on this commit message I would choose something that uses more like an "MVC pattern" and also would try to avoid a DB scheme here.
> 
> We don't need to discuss the choice for SQLite here, but my suggestion would be to split this into a proper package (for instance `avocado.core.persistence.backends.sqlite`) with a very basic interface, so if we decide to change this later, we could do it easily. Also for better organization, avoiding specific SQLite code inside the `requirements` module.

You're supposing that we will want to support multiple backends?  IMO there's no proof that we need multiple backends to solve the same problem... it would only increase our maintenance effort at this stage.

Then, we are free to change this interface and implementation at any time: it's pretty much an internal interface, it's "opt-in" (that is, not required that anyone uses) and it's limited to 2 simple functions.

File: avocado/plugins/spawners/podman.py (hunk 3 above)
Comment (clebergnu) [reply avocado/plugins/spawners/podman.py:173]: Right, the composition is intentional.  The requirements check may also add to the `task_info.status` about what happened there.

File: avocado/core/magic.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/magic.py:17]: OK, so let's drop this, and if the need arises, we can *add* a more specific "fingerprint" print/check.

File: avocado/core/status/repo.py [hunk 8]
```diff
@@ -45,6 +45,10 @@ def get_task_data(self, task_id):
         """Returns all data on a given task, by its ID."""
         return self._all_data.get(task_id)
 
+    def get_latest_task_data(self, task_id):
+        """Returns the latest data on a given task, by its ID."""
+        return self._all_data.get(task_id)[-1]
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/core/status/repo.py:51]: Which will raise an `IndexError` already... so are you suggesting raising a new and specific exception to be handled by the caller?  I think, based on the API, that's better to return `None` to the caller, telling that "there's no latest task data".  What do you think?

File: selftests/unit/test_task_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply selftests/unit/test_task_info.py:15]: > > IMO this usage is very strange. `task.task_info.status`, instead of a direct `task.status`. I understand that at some point tasks will not have all information, but as I said before in a previous review, from the OOP perspective there is no problem with this being part of `Task`. I can't see the reason for this separation.
> 
> Do you think a task should know that it's supposed to be started by a spawner? And that it should have a place for the spawner's own handling? I find this very wrong. You mention the "OOP perspective", but composition is widely used in OOP.
> 

I can't see why not. And you find this very wrong. Yes, the composition is widely used but I don't see the need here, at least how it is described in this PR. I'm also not saying that the current model is wrong from the OOP perspective.

When you do things like:

```python
is_task_alive(task_info)
spawn_task(task_info)
```
IMO, would be better if we had:

```python
is_task_alive(task)
spawn_task(task)
```

> > We are creating a new class because only 4 attributes that are related to a Task.
> 
> Right, and I think they're in the right place. If you look at how a task _can be run_, say, `Task.run()` or `avocado-runner-* task-run ...` it needs none of the attributes. So they're realted to a Task, but they're not part of a Task.

When you describe all 4 items (`status`, `execution_timeout`, `spawner_handle` and `spawning_result`) you mention the word "task" in the docstrings. So yes, I can see this as `task.status`, `task.execution_timeout`. Without a problem.

File: avocado/plugins/nrun.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/plugins/nrun.py]: `pending_tasks`, give us the impression that there are "TASKS" there.

File: avocado/core/requirements.py (hunk 2 above)
Comment (beraldoleal) [reply avocado/core/requirements.py:50]: Even if we don't have multiple backends, IMO is valid the split for better organization.

File: avocado/core/status/repo.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/core/status/repo.py:51]: A `TypeError` could be raised here, if .get() returns `None`.

[the remark this answers is not in the document]
Comment (pevogam) [reply avocado/core/status/repo.py:51]: I think `TypeError` will make it less clear way to say that there is no latest task data. Returning `None` could be a natural way and it could even be done without extra `try` clauses like:
```python
task_data = self._all_data.get(task_id)
if task_data is None:
    return None
else:
    return task_data[-1]
```

File: avocado/core/requirements.py (hunk 2 above)
Comment (pevogam) [reply avocado/core/requirements.py:50]: I think splitting it will also make it more readable as the current code will be abstracted away from such implementation details.

File: avocado/core/status/repo.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply avocado/core/status/repo.py:51]: I'm ok with that.

File: selftests/unit/test_task_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (clebergnu) [reply selftests/unit/test_task_info.py:15]: Well, there's certainly an impasse here, and that's fine because "wisdom lives in the multitude of opinion".  While `is_task_alive(task)` is certainly an improvement over `is_task_alive(task_info)` for readability purposes, the ownership of the extra data is misplaced.

If we combine `TaskInfo` attributes within a `Task`, the following would happen:
 
```python
task.status = 'CHECKED REQUIREMENTS OK on podman image "foobar"'
PodmanSpawner.spawn_task(task)
cmd = podman_bin + "run" + image + task.get_args() # serialization into args
```

Within the container, the task:

```python
task = Task.from_commandline_args() # deserialization
# Task is alive and running
task.status == None # Nothing here?
task.execution_timeout == None # can I set it? no, not a test timeout, not handled by the Task itself... 
task.spawner_handle == None # No one spawned me?
```

And, if one is to write a `Task` implementation (as part of a `avocado-runner-*`) in any other language (which is absolutely a goal here), these extra attributes at the reference (Python) implementation would only to the development overhead or confusion, as it would make no sense it all within that context.

File: avocado/core/requirements.py (hunk 2 above)
Comment (clebergnu) [reply avocado/core/requirements.py:50]: So this is what the split *change* would look like:

https://github.com/clebergnu/avocado/commit/34d40f800832012c0009ce372042091d4a2f66aa

And this is what the entirety of the resulting `avocado.core.requirements` would look like (minus comments):

```
from .requirementbackend import (set_requirement_on_cache,
                                 get_requirement_on_cache)
```

I can go any way, but this looks like over engineering at this point, and IMO, makes it more confusing to follow the code.

Do you guys think it's worth it?

Comment (pevogam) [reply avocado/core/requirements.py:50]: I am not sure about the merged `requirementbackend` module name (probably best to have it as `requirements_backend.sqlite` or something) but I am still for this pattern. My main motivation is that this is an area of the code that will no doubt grow a lot when the we discuss and establish more of the requirements resolver blueprint and even now we could be fairly certain that the requirements definitions and checking/validation would need more implementations to come.

So indeed, it could look a bit like over-engineering at the moment to have a list with one item of supported backends but (1) it will quite certainly be needed very soon and (2) it is a good pattern by itself and even with one item we can see that your suggested changes were mostly separating the code into another module and leaving the first module easier to read and abstracted from the second.

Comment (clebergnu) [reply avocado/core/requirements.py:50]: About the naming, I know @beraldoleal doesn't like names with underscores (and he's got a point because Python suggests against it).  So, I'm finding it hard to name it because something like `requirements.backends.sqlite` would imply that we have and select one of the available backends, while we dont...

Suggestions on top of this comment?

Comment (pevogam) [reply avocado/core/requirements.py:50]: Good point. The `requirements.backends.sqlite` is clean of underscores so I guess the only trouble with it is that there are no modules other than `sqlite` which is more or less the single item list argument we mentioned above but in terms of files. In this case I would still go for it, and we will gradually populate the `requirements.backends` subpackage with more options.

The module `requirements` itself could just have an attribute for the selected backend:
```python
# inside the requirements.py:

from .backends import *

BACKEND = sqlite

# more code here
```

Forgive me if this example is too simplistic.

Comment (clebergnu) [reply avocado/core/requirements.py:50]: I can go the `requirements.backends.sqlite` route... but I don't really see more than *one* backend being supported... I can think of a use case right now.  This is what is hurting me.

Maybe, if we switch the `backend` for `cache`, it could simply be, at this time `avocado.core.requirements.cache`, with a single implementation at this time.  *IF* we get a second implementation, we introduce the module `backends` namespace.

How does that sound?

Comment (pevogam) [reply avocado/core/requirements.py:50]: It is not as explicit as `sqlite` but `cache` sounds reasonable middle ground considering the scarcity of backends and backend types we have at the moment. I have an idea about a requirements backend to suggest but it is in its early "drafts" and I would rather first focus on suggesting spawners and models of isolation. Some requirements backends might also end up entangled with some specific spawners so I am fine with your current choice before I could say more :-1:

Comment (pevogam) [reply avocado/core/requirements.py:50]: > I can go the `requirements.backends.sqlite` route... but I don't really see more than _one_ backend being supported... I can think of a use case right now. This is what is hurting me.

Also what matters is that the `requirements` module is left free of assumptions about cache and other implementations so this should be enough for the time being.

File: selftests/unit/test_task_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply selftests/unit/test_task_info.py:15]: Hi @clebergnu First of all: I don't want to be pokey here :) This just sounds strange to me and I'm trying to make this consistent. We are losing a lot with readability and maintainability, IMO. but feel free to ignore and move on, we can discuss this after LTS. Said that...
 
I'm not seeing the big picture here. But one problem that I was expecting to be fixed was this task des/ser from the command line. I thought that we had discussed this before, and we agreed that this was not so good. Probably I didn't understand properly.

IIUC, you are saying that we need a "composition" because we don't need a few attributes when deserializing? This doesn't feel the right argument for needing a composition. What if we use inheritance here?

BasicTask
CompleteTask(BasicTask):
   + plus those extra args?

In that way, we can have `task.status` (where `task` is a `CompleteTask()`) (The names are just suggestions)

File: avocado/core/requirements.py (hunk 2 above)
Comment (beraldoleal) [reply avocado/core/requirements.py:50]: > So, I'm finding it hard to name it because something like requirements.backends.sqlite would imply that we have and select one of the available backends, while we don't...

```python
from avocado.core.backends.sqlite import (methods here)
``` 

Probably we are going to disagree here too. :) lol But I don't think it is a problem to have this organization even with only one backend supported. But if you don't like that, maybe we could have:

```python
from avocado.utils.sqlite import (methods here)
```

or 

```python
from avocado.core.sqlite import (methods here)
```

where methods are not related to "requirements", would be something like query, save, create_table, drop, delete, ....

File: selftests/unit/test_task_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (clebergnu) [reply selftests/unit/test_task_info.py:15]: > Hi @clebergnu First of all: I don't want to be pokey here :) This just sounds strange to me and I'm trying to make this consistent. We are losing a lot with readability and maintainability, IMO. but feel free to ignore and move on, we can discuss this after LTS. Said that...
> 
> I'm not seeing the big picture here. But one problem that I was expecting to be fixed was this task des/ser from the command line. I thought that we had discussed this before, and we agreed that this was not so good. Probably I didn't understand properly.
> 

What do you mean by fixed? I don't really remember what the core issue is.

> IIUC, you are saying that we need a "composition" because we don't need a few attributes when deserializing? This doesn't feel the right argument for needing a composition. What if we use inheritance here?
> 
> class BasicTask:
> class CompleteTask(BasicTask):
> 
> * plus those extra args?
> 
> In that way, we can have `task.status` (where `task` is a `CompleteTask()`) (The names are just suggestions)

It doesn't look like we'll agree here.  I find this approach much worse for reasons explained [here](https://github.com/avocado-framework/avocado/pull/4128#discussion_r475724946) and also for the fact that `Tasks` are created much earlier, during the suite creation.  The old `avocado nlist` command, for instance, had a `--write-recipes-to-directory` which would persist those Tasks.  And the current `avocado list --resolver` won't ever attempt to deal with their execution (and could/should soon have a similar `--write-recipes-to-directory`). 

So we either favor creating the `CompleteTask` in most places or we do something like `CompleteTask.from_basic_task(task)`.

So, to summarize, I'm not convinced inheritance is the best solution here.

File: avocado/core/requirements.py (hunk 2 above)
Comment (clebergnu) [reply avocado/core/requirements.py:50]: Well, the problem is that we currently don't have (and this don't need) utilities for query, save, create_table, drop, delete. Python's dbapi and `sqlite3` module is rather straightforward to use.

File: selftests/unit/test_task_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply selftests/unit/test_task_info.py:15]: I see, yes, the inheritance was just a suggestion to this current approach. No worries.

For me the strange part is:

"Where should I get 'Task information'?" In my mind, the answer is straightforward: All information related to a task should be inside the Task. Having another "TaskInfo" sounds very strange. And it is even strange when I do thinks like `is_task_alive(task_info)` or `spawn_task(task_info)`, as I said before.

But ok, probably we are not going to agree here. No worries, life continues. ;)

[the remark this answers is not in the document]
Comment (pevogam) [reply selftests/unit/test_task_info.py:15]: I think this will be easier to resolve in the future where the actual use cases establish what is more appropriate way of structuring the data. So as it might be unclear for all sides, let's just give it some time and see where the implementation and its usage bring us.

[the remark this answers is not in the document]
Comment (clebergnu) [reply selftests/unit/test_task_info.py:15]: So the latest idea/proposal is to rename `TaskInfo` to `RuntimeTask`, so that the current `Task` can be continued to be used as is, and something like this:

```python
task_to_be_executed = RuntimeTask(task)
```

How does it sound?

[the remark this answers is not in the document]
Comment (beraldoleal) [reply selftests/unit/test_task_info.py:15]: I think that is better. :) Thanks a lot.

File: avocado/plugins/nrun.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (clebergnu) [reply avocado/plugins/nrun.py]: I'm assuming that with the rename to `RuntimeTask`, this is less ambiguous now.

File: avocado/core/requirements.py (hunk 2 above)
Comment (willianrampazzo) [reply avocado/core/requirements.py:50]: I'm also on the side that any direct interaction with a database should not be in this module. The methods responsible to interact with the `cache`/backend should be abstract enough to do not break this module in case we change the backend.

Comment (clebergnu) [reply avocado/core/requirements.py:50]: OK, I've reorganized the module structure on v2.

File: avocado/core/status/repo.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (willianrampazzo) [reply avocado/core/status/repo.py:51]: > I think `TypeError` will make it less clear way to say that there is no latest task data. Returning `None` could be a natural way and it could even be done without extra `try` clauses like:
> 
> ```python
> task_data = self._all_data.get(task_id)
> if task_data is None:
>     return None
> else:
>     return task_data[-1]
> ```

Just a small optimization as there is no need to have an `else` here:

 ```python
 task_data = self._all_data.get(task_id)
 if task_data is None:
     return None
 return task_data[-1]
 ```

File: selftests/unit/test_task_info.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (willianrampazzo) [reply selftests/unit/test_task_info.py:15]: If I recall correctly, we discussed that the status of a task is not part of the task, but part of the state where the task is at one specific moment. I think the confusion here is more related to the structure naming than anything else.

[the remark this answers is not in the document]
Comment (clebergnu) [reply selftests/unit/test_task_info.py:15]: Yes, that's true.  Seems to be solved by the rename! :)

**Comment by clebergnu** [comment]
Thanks for the review @pevogam @beraldoleal @willianrampazzo , I'm sending a v2 now.

**Comment by clebergnu** [comment]
Closing in favor of #4129
